dgraph/cmd/graphql/run.go

```diff
@@ -0,0 +1,301 @@
+)
+
+type options struct {
+	schemaFile     string
+	alpha          string
+	useCompression bool
+}
+
+type graphqlSchema struct {
+	Type   string    `json:"dgraph.type,omitempty"`
+	Schema string    `json:"dgraph.graphql.schema.schema,omitempty"`
+	Date   time.Time `json:"dgraph.graphql.schema.date,omitempty"`
+}
+
+type graphqlSchemas struct {
+	Schemas []graphqlSchema `json:"graphqlSchemas,omitempty"`
+}
+
+var opt options
+
+var GraphQL x.SubCommand
+
+func init() {
+	GraphQL.Cmd = &cobra.Command{
+		Use:   "graphql",
+		Short: "Run the Dgraph GraphQL tool",
+		Run: func(cmd *cobra.Command, args []string) {
+			defer x.StartProfile(GraphQL.Conf).Stop()
+			run()
+		},
+	}
+
+	GraphQL.EnvPrefix = "DGRAPH_GRAPHQL"
+
+	flags := GraphQL.Cmd.Flags()
+	flags.StringP("alpha", "a", "127.0.0.1:9080",
+		"Comma-separated list of Dgraph alpha gRPC server addresses")
+	flags.BoolP("use_compression", "C", false,
+		"Enable compression on connection to alpha server")
+	flags.StringP("schema", "s", "schema.graphql",
+		"Location of GraphQL schema file")
+
+	cmdInit := &cobra.Command{
+		Use:   "init",
+		Short: "Initializes Dgraph for a GraphQL schema",
+		Args:  cobra.NoArgs,
+		Run: func(cmd *cobra.Command, args []string) {
+			defer x.StartProfile(GraphQL.Conf).Stop()
+			if err := initDgraph(); err != nil {
+				fmt.Printf("Error : %s\n", err)
+				os.Exit(1)
+			}
+		},
+	}
+	cmdInit.Flags().AddFlag(GraphQL.Cmd.Flag("alpha"))
+	cmdInit.Flags().AddFlag(GraphQL.Cmd.Flag("schema"))
+	GraphQL.Cmd.AddCommand(cmdInit)
+
+	// TLS configuration
+	x.RegisterClientTLSFlags(flags)
+}
+
+func run() {
+	x.PrintVersion()
+	opt = options{
+		schemaFile:     GraphQL.Conf.GetString("schema"),
+		alpha:          GraphQL.Conf.GetString("alpha"),
+		useCompression: GraphQL.Conf.GetBool("use_compression"),
+	}
+
+	glog.Infof("Bringing up GraphQL API for Dgraph at %s\n", opt.alpha)
+
+	dgraphClient, disconnect := connect()
+	defer disconnect()
+
+	q := `query {
+		graphqlSchemas(func: type(dgraph.graphql.schema)) {
+			dgraph.graphql.schema.schema
+			dgraph.graphql.schema.date
+		}
+	}`
+
+	ctx := context.Background()
+	resp, err := dgraphClient.NewTxn().Query(ctx, q)
+	x.Checkf(err, "Error querying GraphQL schema from Dgraph")
+
+	var schemas graphqlSchemas
+	err = json.Unmarshal(resp.Json, &schemas)
+	x.Checkf(err, "Error reading GraphQL schema")
+
+	doc, gqlErr := parser.ParseSchema(&ast.Source{Input: string(schemas.Schemas[0].Schema)})
+	if gqlErr != nil {
+		x.Checkf(gqlErr, "Error parsing GraphQL schema")
+	}
+
+	schema, gqlErr := validator.ValidateSchemaDocument(doc)
+	if gqlErr != nil {
+		x.Checkf(gqlErr, "Error validating GraphQL schema")
+	}
+
+	handler := &graphqlHandler{
+		dgraphClient: dgraphClient,
+		schema:       schema,
+	}
+
+	http.Handle("/graphql", handler)
+
+	// the ports and urls etc that the endpoint serves should be input options
+	glog.Fatal(http.ListenAndServe(":8765", nil))
```

Comment:
We should avoid calling Fatal going forward in the new code. Panic works better given that it executes all the defer calls before killing the process.